bee-common/bee-common/src/packable.rs

```diff
@@ -0,0 +1,75 @@
+// Copyright 2020 IOTA Stiftung
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+// an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and limitations under the License.
+
+//! A module that provides a `Packable` trait to serialize and deserialize types.
+
+pub use std::io::{Read, Write};
+
+/// A trait to pack and unpack types to and from bytes.
+pub trait Packable {
+    /// Associated error type.
+    type Error;
+
+    /// Returns the length of the packed bytes.
+    fn packed_len(&self) -> usize;
+
+    /// Packs the instance to bytes and writes them to the passed writer.
+    fn pack<W: Write>(&self, writer: &mut W) -> Result<(), Self::Error>;
+
+    /// Packs the instance to bytes and writes them to a newly allocated vector.
+    fn pack_new(&self) -> Result<Vec<u8>, Self::Error> {
+        let mut bytes = Vec::with_capacity(self.packed_len());
+        self.pack(&mut bytes)?;
+
+        Ok(bytes)
+    }
+
+    /// Reads bytes from the passed reader and unpacks them into an instance.
+    fn unpack<R: Read + ?Sized>(reader: &mut R) -> Result<Self, Self::Error>
+    where
+        Self: Sized;
+}
+
+macro_rules! impl_packable_for_num {
+    ($ty:ident) => {
+        impl Packable for $ty {
+            type Error = std::io::Error;
+
+            fn packed_len(&self) -> usize {
```

Comment:
This doesn't look correct. `size_of` doesn't include the size of associated buffers or things pointed to by this type.